src/hotspot/cpu/x86/x86.ad

```diff
@@ -1835,12 +1838,104 @@ const bool Matcher::match_rule_supported_vector(int opcode, int vlen, BasicType
         return false;
       }
       break;
+    case Op_MaskAll:
+      if(!VM_Version::supports_avx512vlbw()) {
+        return false;
+      }
+      break;
   }
   return true;  // Per default match rules are supported.
 }
 
 const bool Matcher::match_rule_supported_vector_masked(int opcode, int vlen, BasicType bt) {
-  return false;
+  // Needed for loadmask pattern which populates opmask register
+  // consumed by masked instructions.
+  if (!VM_Version::supports_avx512bw()) {
+    return false;
+  }
+  int size_in_bits = vlen * type2aelembytes(bt) * BitsPerByte;
+  if ((size_in_bits != 512) && !VM_Version::supports_avx512vl()) {
+     return false; // Implementation limitation
+  }
+  switch(opcode) {
+    // Unary masked operations
+    case Op_AbsVI:
+      return true;
+
+    // Ternary masked operations
+    case Op_FmaVF:
+    case Op_FmaVD:
+      return true;
+
+    // Binary masked operations
+    case Op_AddVB:
+    case Op_AddVS:
+    case Op_SubVB:
+    case Op_SubVS:
+    case Op_MulVS:
+    case Op_LShiftVS:
+    case Op_RShiftVS:
+    case Op_URShiftVS:
+      assert(VM_Version::supports_avx512bw(), "");
+      assert(size_in_bits == 512 || VM_Version::supports_avx512vl(), "");
+      return true;
+
+    case Op_MulVL:
+      assert(size_in_bits == 512 || VM_Version::supports_avx512vl(), "");
+      if (!VM_Version::supports_avx512dq()) {
+        return false;  // Implementation limitation
+      }
+      return true;
+
+    case Op_AddVI:
+    case Op_AddVL:
+    case Op_AddVF:
+    case Op_AddVD:
+    case Op_SubVI:
+    case Op_SubVL:
+    case Op_SubVF:
+    case Op_SubVD:
+    case Op_MulVI:
+    case Op_MulVF:
+    case Op_MulVD:
+    case Op_DivVF:
+    case Op_DivVD:
+    case Op_LShiftVI:
+    case Op_LShiftVL:
+    case Op_RShiftVI:
+    case Op_RShiftVL:
+    case Op_URShiftVI:
+    case Op_URShiftVL:
+    case Op_LoadVectorMasked:
+    case Op_StoreVectorMasked:
+    case Op_LoadVectorGatherMasked:
+    case Op_StoreVectorScatterMasked:
+      return true;
+
+    case Op_MaxV:
+    case Op_MinV:
+      if (is_floating_point_type(bt)) {
+        return false; // Implementation limitation
+      }
+      return true;
+
+    case Op_VectorRearrange:
+      assert(bt != T_SHORT || VM_Version::supports_avx512bw(), "");
+      if (bt == T_BYTE && VM_Version::supports_avx512_vbmi()) {
+        return false; // Implementation limitation
+      }
+      return true;
```

Comment:
I guess the masked `VectorRearrange` is not generated currently, is it? So if AVX-512 supports the masked operations, will you add modify the API and intrinsics side for it in future?

Reply:
Yes, rearrange Template needs to accommodate extra mask argument, I added the backend changes in anticipation.

Reply:
OK, thank you for that. Since SVE doesn't support this masked operation, we will block it in SVE's backend.